File: app/lib/models/history.dart
class HistoryProject {
  final String projectDir;
  final String cwd;
  final String? repo;
  final String label;
  final int sessionCount;
  final String lastActivity;
  final String? nodeId;
  final String? nodeLabel;

  const HistoryProject({
    required this.projectDir,
    required this.cwd,
    this.repo,
    required this.label,
    required this.sessionCount,
    required this.lastActivity,
    this.nodeId,
    this.nodeLabel,
  });

  factory HistoryProject.fromJson(Map<String, dynamic> j) => HistoryProject(
        projectDir: j['project_dir'] as String? ?? '',
        cwd: j['cwd'] as String? ?? '',
        repo: j['repo'] as String?,
        label: j['label'] as String? ?? '',
        sessionCount: (j['session_count'] as num?)?.toInt() ?? 0,
        lastActivity: j['last_activity'] as String? ?? '',
        nodeId: j['node_id'] as String?,
        nodeLabel: j['node_label'] as String?,
      );
}

class HistorySession {
  final String sessionId;
  final String? title;
  final String? firstMessage;
  final String transcriptPath;
  final String? modelName;
  final String? modelColor;
  final String lastActivity;
  final int tokens;
  final int turnCount;
  final int durationMs;
  final String? nodeId;
  final String? nodeLabel;

  const HistorySession({
    required this.sessionId,
    this.title,
    this.firstMessage,
    required this.transcriptPath,
    this.modelName,
    this.modelColor,
    required this.lastActivity,
    required this.tokens,
    required this.turnCount,
    required this.durationMs,
    this.nodeId,
    this.nodeLabel,
  });

  factory HistorySession.fromJson(Map<String, dynamic> j) => HistorySession(
        sessionId: j['session_id'] as String? ?? '',
        title: j['title'] as String?,
        firstMessage: j['first_message'] as String?,
        transcriptPath: j['transcript_path'] as String? ?? '',
        modelName: j['model_name'] as String?,
        modelColor: j['model_color'] as String?,
        lastActivity: j['last_activity'] as String? ?? '',
        tokens: (j['tokens'] as num?)?.toInt() ?? 0,
        turnCount: (j['turn_count'] as num?)?.toInt() ?? 0,
        durationMs: (j['duration_ms'] as num?)?.toInt() ?? 0,
        nodeId: j['node_id'] as String?,
        nodeLabel: j['node_label'] as String?,
      );

  /// The label shown wherever a history session needs a title: title, else the
  /// first message, else the session id.
  String get displayTitle => (title?.isNotEmpty ?? false)
      ? title!
      : (firstMessage?.isNotEmpty ?? false)
          ? firstMessage!
          : sessionId;
}

class HistorySessionPage {
  final List<HistorySession> items;
  final bool hasMore;

  const HistorySessionPage({
    required this.items,
    required this.hasMore,
  });

  factory HistorySessionPage.fromJson(Map<String, dynamic> j) =>
      HistorySessionPage(
        items: (j['items'] as List?)
                ?.map((e) => HistorySession.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        hasMore: j['has_more'] as bool? ?? false,
      );
}

File: internal/session/history.go
package session

// HistoryProject is one Claude Code project directory discovered on disk, for the
// read-only History view. NodeID/NodeLabel are empty node-side and stamped by
// the gateway when aggregating across machines.
type HistoryProject struct {
	ProjectDir   string `json:"project_dir"`          // routing key for its sessions (node-local path)
	Cwd          string `json:"cwd"`                  // working directory of the project
	Repo         string `json:"repo,omitempty"`       // git repo basename, when inside a repo
	Label        string `json:"label"`                // display name (repo, else cwd basename)
	SessionCount int    `json:"session_count"`        // number of transcripts in the project
	LastActivity string `json:"last_activity"`        // RFC3339 (UTC) newest transcript mod time
	NodeID       string `json:"node_id,omitempty"`    // origin machine (gateway-stamped)
	NodeLabel    string `json:"node_label,omitempty"` // origin machine label (gateway-stamped)
}

// HistorySession is one past session within a project, for the History session list.
// It carries no live/tmux state; TranscriptPath + NodeID address its transcript.
type HistorySession struct {
	SessionID      string `json:"session_id"`
	Title          string `json:"title,omitempty"`         // custom/AI title, when present
	FirstMessage   string `json:"first_message,omitempty"` // first user message (title fallback)
	TranscriptPath string `json:"transcript_path"`         // node-local path; routing key for the view
	ModelName      string `json:"model_name,omitempty"`
	ModelColor     string `json:"model_color,omitempty"`
	LastActivity   string `json:"last_activity"` // RFC3339 (UTC) transcript mod time
	Tokens         int    `json:"tokens,omitempty"`
	TurnCount      int    `json:"turn_count,omitempty"`
	DurationMs     int64  `json:"duration_ms,omitempty"`
	NodeID         string `json:"node_id,omitempty"`
	NodeLabel      string `json:"node_label,omitempty"`
}

// HistorySessionPage is a window of a project's sessions, newest-first, with a flag
// indicating whether older sessions remain beyond this page.
type HistorySessionPage struct {
	Items   []HistorySession `json:"items"`
	HasMore bool             `json:"has_more"`
}
